Allows substring matching with 'mason test'  inside mason package

rajshekhar7 opened the pull request:
This PR supports filtering similar to `cargo test`.
 
Now, `mason test test/test1.chpl test/test2.chpl`    can also be run using 

```
mason test test1 test2
# or
mason test test
# or
mason test t
```

This was mentioned in #15068

rajshekhar7 commented:
@ben-albrecht , Please take a look at this.

ben-albrecht commented:
Also, we need a CLA, before we can merge this PR.

rajshekhar7 commented:
@ben-albrecht  Can you give me an example of how arguments for compopts are specified with mason test?

ben-albrecht commented:
>  Can you give me an example of how arguments for compopts are specified with mason test?

```
mason test -- --fast
mason test -- --baseline
```

rajshekhar7 commented:
@ben-albrecht  

If we have some tests listed in **Mason.toml**, should the substring matching be done only with those files or with all the files in the test directory?

Right now, if we have some tests listed in Mason.toml and we run `mason test test/test1.chpl`, it ignores all the files listed in Mason.toml and runs the tests only on `test/test1.chpl`

ben-albrecht commented:
> If we have some tests listed in Mason.toml, should the substring matching be done only with those files or with all the files in the test directory?

> Right now, if we have some tests listed in Mason.toml and we run mason test test/test1.chpl, it ignores all the files listed in Mason.toml and runs the tests only on test/test1.chpl


@rajshekhar7 - that's a good question. I would advocate keeping the current behavior of ignoring `Mason.toml` tests. I think of the `Mason.toml` tests as the standard test suite of a package, where there may be other tests in `test/` that a user may want to run manually.

This decision may be revisited in the future though.

rajshekhar7 commented:
@ben-albrecht 

- compopts has been separated from the file search string at the argument-parsing level. Although if a substring is not matched with any of the files, it will be sent to compopts which will throw an unrecognized suffix error.

I've added test for this feature, ran testing and updated the docs. Also, I've sent the CLA.

Please review the changes.

ben-albrecht commented on test/mason/masonTestSubString/mason-test-sub-string.chpl:
```diff
@@ -0,0 +1,9 @@
+use MasonTest;
+
```
We don't need all these newlines.

rajshekhar7 commented:
@ben-albrecht 

> I noticed test/library/packages/UnitTest/Launcher_Test.chpl does not pass with these changes.

I've fixed this error.

> Does this substring work with mason test outside of a mason package too?

I've added the support to use substring outside of the mason package with the above fix and ran testing.

Thanks for the documentation suggestion!

Please review the changes.

rajshekhar7 commented:
@ben-albrecht 

Fixed the error and ran the tests. Please have a look.

rajshekhar7 commented:
@ben-albrecht - I've updated this PR for 0-based indexing and ran the tests. Please have a look.

rajshekhar7 commented:
@ben-albrecht - Can you take a look at this PR. Are there any additional changes to be made?
